Userland/Services/WebContent/RenderLoopExecutor.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2024, Aliaksandr Kalenik <kalenik.aliaksandr@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/EventLoop.h>
+#include <LibGfx/ShareableBitmap.h>
+#include <LibWeb/Painting/PaintingCommandExecutorCPU.h>
+#include <WebContent/RenderLoopExecutor.h>
+
+#ifdef HAS_ACCELERATED_GRAPHICS
+#    include <LibWeb/Painting/PaintingCommandExecutorGPU.h>
+#endif
+
+namespace WebContent {
+
+RenderLoopExecutor::RenderLoopExecutor(bool use_gpu_painter)
+    : m_main_thread_event_loop(Core::EventLoop::current())
+    , m_use_gpu_painter(use_gpu_painter)
+{
+}
+
+RenderLoopExecutor::~RenderLoopExecutor()
+{
+    m_exit = true;
+    m_paint_task_ready_wake_condition.signal();
+}
+
+void RenderLoopExecutor::paint(NonnullOwnPtr<Web::Painting::RecordingPainter> recording_painter, Function<void(i32)> on_completion)
+{
+    if (!m_backing_stores.back_bitmap)
+        return;
+
+    enqueue_paint_task([this, recording_painter = move(recording_painter), on_completion = move(on_completion)]() mutable {
+        m_backing_stores_mutex.lock();
+        auto& bitmap = *m_backing_stores.back_bitmap;
+
+        if (m_use_gpu_painter) {
+#ifdef HAS_ACCELERATED_GRAPHICS
+            Web::Painting::PaintingCommandExecutorGPU painting_command_executor(*m_accelerated_graphics_context, bitmap);
+            recording_painter->execute(painting_command_executor);
+#else
+            static bool has_warned_about_configuration = false;
+
+            if (!has_warned_about_configuration) {
+                warnln("\033[31;1mConfigured to use GPU painter, but current platform does not have accelerated graphics\033[0m");
+                has_warned_about_configuration = true;
+            }
+#endif
+        } else {
+            Web::Painting::PaintingCommandExecutorCPU painting_command_executor(bitmap);
+            recording_painter->execute(painting_command_executor);
+        }
+
+        swap(m_backing_stores.front_bitmap, m_backing_stores.back_bitmap);
+        swap(m_backing_stores.front_bitmap_id, m_backing_stores.back_bitmap_id);
+        auto front_bitmap_id = m_backing_stores.front_bitmap_id;
+        m_backing_stores_mutex.unlock();
+
+        m_main_thread_event_loop.deferred_invoke([on_completion = move(on_completion), front_bitmap_id]() {
+            on_completion(front_bitmap_id);
+        });
+
+        m_ready_to_paint_mutex.lock();
+        m_ready_to_paint_wake_condition.wait();
+        m_ready_to_paint_mutex.unlock();
+    });
+}
+
+void RenderLoopExecutor::ready_to_paint()
+{
+    m_ready_to_paint_mutex.lock();
+    m_ready_to_paint_wake_condition.signal();
+    m_ready_to_paint_mutex.unlock();
+}
+
+void RenderLoopExecutor::start()
+{
+    m_thread = Threading::Thread::construct([this]() mutable -> intptr_t {
```

Comment:
`intptr_t` feels not very serenity-y, is this mandated by our thread api, or can this be something nicer?

Reply:
That's mandated by LibThreading. to mimic the void* of pthread_create()

Reply:
yes, it is mandated by out thread api:
`static NonnullRefPtr<Thread> construct(Function<intptr_t()> action, StringView thread_name = {})`